common/setups/rasr/util/nn.py

```diff
@@ -198,52 +195,113 @@ def get_data_dict(self):
         }
 
 
+# Attribute names are invalid identifiers, therefore use old syntax
+SearchParameters = TypedDict(
+    "SearchParameters",
+    {
+        "beam-pruning": float,
+        "beam-pruning-limit": float,
+        "lm-state-pruning": Optional[float],
+        "word-end-pruning": float,
+        "word-end-pruning-limit": float,
+    },
+)
+
+
+class LookaheadOptions(TypedDict):
+    cache_high: Optional[int]
+    cache_low: Optional[int]
+    history_limit: Optional[int]
+    laziness: Optional[int]
+    minimum_representation: Optional[int]
+    tree_cutoff: Optional[int]
+
+
+class LatticeToCtmArgs(TypedDict):
+    best_path_algo: Optional[str]
+    encoding: Optional[str]
+    extra_config: Optional[Any]
+    extra_post_config: Optional[Any]
+    fill_empty_segments: Optional[bool]
+
+
+class NnRecogArgs(TypedDict):
+    acoustic_mixture_path: Optional[tk.Path]
+    checkpoints: Optional[Dict[int, returnn.Checkpoint]]
+    create_lattice: Optional[bool]
+    epochs: Optional[List[int]]
+    eval_best_in_lattice: Optional[bool]
+    eval_single_best: Optional[bool]
+    feature_flow_key: str
+    lattice_to_ctm_kwargs: Optional[LatticeToCtmArgs]
+    lm_lookahead: bool
+    lm_scales: List[float]
+    lookahead_options: Optional[LookaheadOptions]
+    mem: int
+    name: str
+    optimize_am_lm_scale: bool
+    parallelize_conversion: Optional[bool]
+    prior_scales: List[float]
+    pronunciation_scales: List[float]
+    returnn_config: Optional[returnn.ReturnnConfig]
+    rtf: int
+    search_parameters: Optional[SearchParameters]
+    use_gpu: Optional[bool]
+
+
+class EpochPartitioning(TypedDict):
+    dev: int
+    train: int
+
+
+class NnTrainingArgs(TypedDict):
+    buffer_size: Optional[int]
+    class_label_file: Optional[tk.Path]
+    cpu_rqmt: Optional[int]
+    device: Optional[str]
+    disregarded_classes: Optional[Any]
+    extra_rasr_config: Optional[RasrConfig]
+    extra_rasr_post_config: Optional[RasrConfig]
+    horovod_num_processes: Optional[int]
+    keep_epochs: Optional[bool]
+    log_verbosity: Optional[int]
+    mem_rqmt: Optional[int]
+    num_classes: int
+    num_epochs: int
+    partition_epochs: Optional[EpochPartitioning]
+    save_interval: Optional[int]
+    time_rqmt: Optional[int]
+    use_python_control: Optional[bool]
+
+
 class HybridArgs:
     def __init__(
         self,
         returnn_training_configs: Dict[str, returnn.ReturnnConfig],
         returnn_recognition_configs: Dict[str, returnn.ReturnnConfig],
-        training_args: Dict[str, Any],
-        recognition_args: Dict[str, Dict[str, Dict]],
-        test_recognition_args: Optional[Dict[str, Dict[str, Dict]]] = None,
+        training_args: Dict[str, NnTrainingArgs],
```

Comment:
This still allows a dict, correct?

Reply:
Type inference might complain (which is what we want, I think), but at runtime, typed dicts are equivalent to regular dicts.

Reply:
This typing is wrong, because it says you want a dict of multiple keyed NnTrainingArgs. I assume you want: `Union[Dict[str, Any], NnTrainingArgs]` so either a str->any dict or the correct typed dict.

Reply:
Oh, good catch. Thanks. I wonder if we want to drop the `Dict[str, Any]` part altogether and have type inference complain or if we want to be more laissez-faire, and optionally support the types there, but also not complain if one doesn't use the typed dicts.